File: Strings/Count Occurrences of a given Word in a 2-d Array/main.py
#User function Template for python3
class Solution:
    def __init__(self):
        self.ans = 0
        
    def dfs(self,mat,i,j,idx,target):
        if(i < 0 or j < 0 or i == len(mat) or j == len(mat[0]) or idx >= len(target) or target[idx] != mat[i][j]):
            return
    
        if(idx == len(target)-1):
            self.ans+=1
            return
        temp = mat[i][j]
        mat[i][j] = '0'
        self.dfs(mat,i+1,j,idx+1,target)
        self.dfs(mat,i,j+1,idx+1,target)
        self.dfs(mat,i-1,j,idx+1,target)
        self.dfs(mat,i,j-1,idx+1,target)
        mat[i][j] = temp
        
        
    def findOccurrence(self,mat,target):
        n = len(mat)
        m = len(mat[0])
        
        for i in range(n):
            for j in range(m):
                self.dfs(mat,i,j,0,target)

        return self.ans
#{ 
 # Driver Code Starts
#Initial Template for Python 3
if __name__ == "__main__":
    t=int(input())
    for _ in range(t):
        line=input().strip().split()
        R=int(line[0])
        C=int(line[1])
        mat=[]
        for _ in range(R):
            mat.append( [x for x in input().strip().split()] )
        target=input()
        obj = Solution()
        print(obj.findOccurrence(mat,target))
        print("~")
# } Driver Code Ends

File: Strings/Count Occurrences of a given Word in a 2-d Array/main.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution {
    public:
        int ans = 0;

        void dfs(vector<vector<char>>& mat, int i, int j, int idx, string& target) {
            if(i < 0 || i == mat.size() || j < 0 || j == mat[0].size() ||
               idx >= target.size() || target[idx] != mat[i][j]) return;

            if(idx == target.size() - 1) {
                ans++;
                return;
            }
            char temp = mat[i][j];
            mat[i][j] = '0';

            dfs(mat, i + 1, j, idx + 1, target);
            dfs(mat, i - 1, j, idx + 1, target);

            dfs(mat, i, j + 1, idx + 1, target);
            dfs(mat, i, j - 1, idx + 1, target);
            mat[i][j] = temp;
        }

        int findOccurrence(vector<vector<char>>& mat, string target) {
            int n = mat.size();
            int m = mat[0].size();
            for(int i = 0; i < n; i++) {
                for(int j = 0; j < m; j++) {
                    dfs(mat, i, j, 0, target);
                }
            }
            return ans;
        }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while(t--) {
        int R, C;
        cin >> R >> C;
        vector<vector<char>> mat(R);
        for(int i = 0; i < R; i++) {
            mat[i].resize(C);
            for(int j = 0; j < C; j++)
                cin >> mat[i][j];
        }
        string target;
        cin >> target;
        Solution obj;
        cout << obj.findOccurrence(mat, target) << endl;

        cout << "~"
             << "\n";
    }
}

// } Driver Code Ends